File: packages/sdk-maui/ScreebHooks.cs
namespace Screeb.Maui;

public class ScreebHooks
{
    /// <summary>Hooks version identifier passed verbatim to the native SDK.</summary>
    public required string Version { get; set; }

    /// <summary>
    /// Hook callbacks: key = hook name (e.g. "onSurveyShowed"),
    /// value = async callback receiving the JSON payload string.
    /// </summary>
    public Dictionary<string, Func<string, Task<object?>>> Callbacks { get; set; } = new();
}

/// <summary>
/// Internal registry mapping UUID → C# callback. Used by platform implementations.
/// Thread-safe via ConcurrentDictionary — callbacks are registered from the main thread
/// but invoked from native Android/iOS callback threads.
/// </summary>
internal static class HooksRegistry
{
    private static readonly System.Collections.Concurrent.ConcurrentDictionary<string, Func<string, Task<object?>>> _registry = new();

    internal static Dictionary<string, string> RegisterHooks(ScreebHooks? hooks)
    {
        var uuidMap = new Dictionary<string, string>();
        if (hooks == null) return uuidMap;
        if (hooks.Version != null)
            uuidMap["version"] = hooks.Version;
        foreach (var (key, callback) in hooks.Callbacks)
        {
            var uuid = Guid.NewGuid().ToString("N") + "_" + key;
            _registry[uuid] = callback;
            uuidMap[key] = uuid;
        }
        return uuidMap;
    }

    internal static Func<string, Task<object?>>? Get(string uuid)
        => _registry.TryGetValue(uuid, out var fn) ? fn : null;

    /// <summary>Removes registered callbacks by UUID.</summary>
    internal static void Unregister(IEnumerable<string> uuids)
    {
        foreach (var uuid in uuids)
            _registry.TryRemove(uuid, out _);
    }

    /// <summary>
    /// Removes all registered callbacks. Call on CloseSdk to prevent unbounded
    /// memory growth across repeated InitSdk/StartSurvey/StartMessage calls.
    /// </summary>
    internal static void UnregisterAll() => _registry.Clear();
}

File: packages/sdk-kmp/src/commonMain/kotlin/app/screeb/sdk/kmp/ScreebHooks.kt
package app.screeb.sdk.kmp

/**
 * Hook callbacks to pass to initSdk, startSurvey, or startMessage.
 * [version] is passed verbatim to the native SDK.
 * [callbacks] maps hook names (e.g. "onSurveyShowed") to suspend lambdas
 * that receive the JSON payload string.
 */
data class ScreebHooks(
    val version: String,
    val callbacks: Map<String, suspend (String) -> Any?> = emptyMap(),
)

/**
 * Internal registry mapping UUID → suspend callback.
 * UUID format: "<hookName>_<random>" to aid debugging.
 * All accesses are performed on the Main dispatcher (enforced by callers via withContext(Dispatchers.Main)),
 * so no additional synchronization is required.
 */
internal object HooksRegistry {
    private val registry = mutableMapOf<String, suspend (String) -> Any?>()

    /** Registers callbacks, returns a map of hookName → UUID for passing to native SDK. */
    fun register(hooks: ScreebHooks?): Map<String, String> {
        if (hooks == null) return emptyMap()
        val uuidMap = mutableMapOf<String, String>()
        uuidMap["version"] = hooks.version
        for ((key, callback) in hooks.callbacks) {
            val uuid = "${key}_${randomUuid()}"
            registry[uuid] = callback
            uuidMap[key] = uuid
        }
        return uuidMap
    }

    fun get(uuid: String): (suspend (String) -> Any?)? = registry[uuid]

    fun clear() = registry.clear()
}

/** Platform-provided UUID. Defined as expect/actual. */
internal expect fun randomUuid(): String
